Fix step calls with no argument

stoeffel opened the pull request:
It also makes `nix develop` work on macos

stoeffel committed:
Make flake work on macos

knuton commented on src/lib/game.ts:
```diff
@@ -212,8 +212,8 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
-  this.step = (now: number) => {
     if (paused) return
+  this.step = (now: number = performance.now()) => {
 
```
Why are the lines swapped?

stoeffel replied:
Wow you are fast. I just saw it as well. the formatting confused me during the commit. I added a commit to just format the whole file, I hope that's okay.

stoeffel replied:
no nevermind, I don't like the format commit. did it without.

stoeffel committed:
Fix step calls without an argument

knuton commented:
Things that would be good to capture in commit message:

- Why is this fix being done (i.e. the function assumes to be called as a callback to rAF, but is also called without args from elsewhere)?
- Why is it OK to use the current time as `now` when resuming from pause, in which case `dt = performance.now() - lastStepAt` might be very large? (Is it OK?)
- Why performance.now (instead of another method to obtain time), and how much checking has been done, that it provides useful values? The [note on MDN is a bit vague](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame#timestamp).

stoeffel committed:
Simplify resuming

THis also removes the global state lastStepAt which makes `step` hard to follow.

stoeffel committed:
Draw doesn't take any arguments

stoeffel commented:
@knuton I removed the `time` vs `runningTime` commit (separate PR during defrag).
I changed the code to never have a need to call `performance.now()` and just rely on passing `lastStepAt` instead of it as a global var.

knuton commented on src/lib/game.ts:
```diff
@@ -212,22 +210,21 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
-  this.step = (now: number) => {
-    if (paused) return
+  this.step = (lastStepAt: number, now: number) => {
+    if (paused) return;
```
Sorry, but no `;` :joy: 

Feeling like a caveman in a codebase without formatter, having this kind of PR comment again.

For some reason everything was reformatted with [`standard`](https://standardjs.com/) [years ago](https://github.com/dividat/skifree.js/commit/ff53336e4634d8de164a48485b69fbe56a951d53).

stoeffel replied:
removed

stoeffel committed:
No semicolon

knuton commented on src/lib/game.ts:
```diff
@@ -184,17 +183,16 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
-    paused = false
-    this.step()
+    paused = false 
+    this.loop()
```
It's hard to stop fixing things with how brittle this game is!

This would do two things:

1. Remove dead code (start/reset)
2. Guard `resume` against starting multiple loops if invoked while game is running

```diff
diff --git i/src/lib/game.ts w/src/lib/game.ts
index d6df641..67874ca 100644
--- i/src/lib/game.ts
+++ w/src/lib/game.ts
@@ -13,7 +13,7 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
   const afterCycleCallbacks: Array<any> = []
 
   let sprites = new Array<Sprite>()
-  let paused = false
+  let paused = true
   let runningTime = 0
   let zoom = config.zoom.max
   let time = 0
@@ -182,17 +182,15 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
     })
   }
 
-  this.start = () => {
-    this.loop()
-  }
-
   this.pause = () => {
     paused = true
   }
 
   this.resume = () => {
-    paused = false 
-    this.loop()
+    if (paused) {
+      paused = false
+      this.loop()
+    }
   }
 
   this.isPaused = () => {
@@ -203,13 +201,6 @@ export function Game (mainCanvas: HTMLCanvasElement, skier: Skier) {
     return runningTime
   }
 
-  this.reset = () => {
-    paused = false
-    sprites = new Array<Sprite>()
-    this.start()
-    runningTime = 0
-  }
-
   this.step = (lastStepAt: number, now: number) => {
     if (paused) return
 
```

stoeffel replied:
looking good to me!

knuton committed:
Guard against multiple resumes
